internal/app/machined/pkg/controllers/block/volume_manager.go

```diff
@@ -0,0 +1,442 @@
+				return volumes.DiskContext{}, err
+			}
+
+			return volumes.DiskContext{
+				Disk:       spec,
+				SystemDisk: systemDisk != nil && d.Metadata().ID() == systemDisk.TypedSpec().DiskID,
+			}, nil
+		})
+		if err != nil {
+			return fmt.Errorf("error mapping disks: %w", err)
+		}
+
+		volumeConfigList, err := safe.ReaderListAll[*block.VolumeConfig](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error fetching volume configurations: %w", err)
+		}
+
+		volumeStatusList, err := safe.ReaderListAll[*block.VolumeStatus](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error fetching volume statuses: %w", err)
+		}
+
+		volumeConfigIDs := xslices.ToSet(safe.ToSlice(volumeConfigList, func(vc *block.VolumeConfig) resource.ID { return vc.Metadata().ID() }))
+
+		volumeStatuses := xslices.ToMap(
+			safe.ToSlice(volumeStatusList, func(vs *block.VolumeStatus) *block.VolumeStatus { return vs }),
+			func(vs *block.VolumeStatus) (resource.ID, *block.VolumeStatusSpec) {
+				return vs.Metadata().ID(), vs.TypedSpec()
+			},
+		)
+
+		if volumeStatuses == nil {
+			volumeStatuses = map[resource.ID]*block.VolumeStatusSpec{}
+		}
+
+		// ensure all volume configs have our finalizers
+		for iter := volumeConfigList.Iterator(); iter.Next(); {
+			vc := iter.Value()
+
+			if vc.Metadata().Phase() != resource.PhaseRunning {
+				continue
+			}
+
+			if vc.Metadata().Finalizers().Has(ctrl.Name()) {
+				continue
+			}
+
+			if err = r.AddFinalizer(ctx, vc.Metadata(), ctrl.Name()); err != nil {
+				return fmt.Errorf("error adding finalizer to volume configuration: %w", err)
+			}
+		}
+
+		// remove statuses for volume configs that no longer exist
+		for id := range volumeStatuses {
+			if _, exists := volumeConfigIDs[id]; !exists {
+				delete(volumeStatuses, id)
+
+				if err := r.Destroy(ctx, block.NewVolumeStatus(block.NamespaceName, id).Metadata()); err != nil {
+					return fmt.Errorf("error destroying volume status: %w", err)
+				}
+			}
+		}
+
+		// fill in statuses for volume configs that don't have a status yet
+		for id := range volumeConfigIDs {
+			if _, exists := volumeStatuses[id]; !exists {
+				volumeStatuses[id] = &block.VolumeStatusSpec{
+					Phase: block.VolumePhaseWaiting,
+				}
+			}
+		}
+
+		volumeConfigs := safe.ToSlice(volumeConfigList, func(vc *block.VolumeConfig) *block.VolumeConfig { return vc })
+
+		// re-sort volume configs by provisioning wave
+		slices.SortStableFunc(volumeConfigs, volumes.CompareVolumeConfigs)
+
+		fullyProvisionedWave := math.MaxInt
+		allClosed := true
+
+		for _, vc := range volumeConfigs {
+			// abort on context cancel, as each volume processing might take a while
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+			}
+
+			volumeStatus := volumeStatuses[vc.Metadata().ID()]
+			volumeLogger := logger.With(zap.String("volume", vc.Metadata().ID()))
+
+			if vc.Metadata().Phase() != resource.PhaseRunning {
+				// [TODO]: handle me later
```

Comment:
anything missing to do here? or future work?

Reply:
future work -- once we support tearing down (wiping) volumes for real